File: API/2015_Color_Schaff_Momentum_Trend_Cycle/CS/ColorSchaffMomentumTrendCycleStrategy.cs
namespace StockSharp.Samples.Strategies;

using System;
using System.Collections.Generic;
using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.Messages;

public class ColorSchaffMomentumTrendCycleStrategy : Strategy
{
	private Momentum _fastMomentum;
	private Momentum _slowMomentum;
	private readonly Queue<decimal> _macdQueue = new(); // recent MACD values
	private readonly Queue<decimal> _stQueue = new();	// intermediate ST values
	private int? _prevColor;	// previous color code
	private decimal _prevStc;	// previous STC value

	private StrategyParam<int> _fastMomentumLength;
	private StrategyParam<int> _slowMomentumLength;
	private StrategyParam<int> _cycle;
	private StrategyParam<int> _highLevel;
	private StrategyParam<int> _lowLevel;
	private StrategyParam<bool> _buyPosOpen;
	private StrategyParam<bool> _sellPosOpen;
	private StrategyParam<bool> _buyPosClose;
	private StrategyParam<bool> _sellPosClose;

	public int FastMomentum
	{
		get => _fastMomentumLength.Value;
		set => _fastMomentumLength.Value = value;
	}

	public int SlowMomentum
	{
		get => _slowMomentumLength.Value;
		set => _slowMomentumLength.Value = value;
	}

	public int Cycle
	{
		get => _cycle.Value;
		set => _cycle.Value = value;
	}

	public int HighLevel
	{
		get => _highLevel.Value;
		set => _highLevel.Value = value;
	}

	public int LowLevel
	{
		get => _lowLevel.Value;
		set => _lowLevel.Value = value;
	}

	public bool BuyPosOpen
	{
		get => _buyPosOpen.Value;
		set => _buyPosOpen.Value = value;
	}

	public bool SellPosOpen
	{
		get => _sellPosOpen.Value;
		set => _sellPosOpen.Value = value;
	}

	public bool BuyPosClose
	{
		get => _buyPosClose.Value;
		set => _buyPosClose.Value = value;
	}

	public bool SellPosClose
	{
		get => _sellPosClose.Value;
		set => _sellPosClose.Value = value;
	}

	public ColorSchaffMomentumTrendCycleStrategy()
	{
		_fastMomentumLength = Param(nameof(FastMomentum), 23).SetDisplay("Fast Momentum", "Fast momentum length", "Indicator");
		_slowMomentumLength = Param(nameof(SlowMomentum), 50).SetDisplay("Slow Momentum", "Slow momentum length", "Indicator");
		_cycle = Param(nameof(Cycle), 10).SetDisplay(nameof(Cycle), "Cycle length", "Indicator");
		_highLevel = Param(nameof(HighLevel), 60).SetDisplay("High Level", "Upper threshold", "Indicator");
		_lowLevel = Param(nameof(LowLevel), -60).SetDisplay("Low Level", "Lower threshold", "Indicator");
		_buyPosOpen = Param(nameof(BuyPosOpen), true).SetDisplay("Enable Long", "Allow long entries", "Trading");
		_sellPosOpen = Param(nameof(SellPosOpen), true).SetDisplay("Enable Short", "Allow short entries", "Trading");
		_buyPosClose = Param(nameof(BuyPosClose), true).SetDisplay("Close Long", "Allow closing long positions", "Trading");
		_sellPosClose = Param(nameof(SellPosClose), true).SetDisplay("Close Short", "Allow closing short positions", "Trading");
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		_macdQueue.Clear();
		_stQueue.Clear();
		_prevColor = null;
		_prevStc = 0m;
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		// create momentum indicators using selected lengths
		_fastMomentum = new Momentum { Length = FastMomentum };
		_slowMomentum = new Momentum { Length = SlowMomentum };

		// subscribe to candle series and start processing
		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(ProcessCandle).Start();
	}

	private void ProcessCandle(ICandleMessage candle)
	{
		if (candle.State != CandleStates.Finished)
			return;

		var price = candle.ClosePrice;
		var fast = _fastMomentum.Process(price, candle.OpenTime, true).ToDecimal();
		var slow = _slowMomentum.Process(price, candle.OpenTime, true).ToDecimal();
		var macd = fast - slow;

		// store MACD and keep limited history
		_macdQueue.Enqueue(macd);
		if (_macdQueue.Count > Cycle)
			_macdQueue.Dequeue();

		// find min/max for MACD window
		var llv = decimal.MaxValue;
		var hhv = decimal.MinValue;
		foreach (var value in _macdQueue)
		{
			if (value < llv)
				llv = value;
			if (value > hhv)
				hhv = value;
		}

		// normalize MACD into ST range
		var st = hhv == llv ? (_stQueue.Count > 0 ? _stQueue.Peek() : 0m) : (macd - llv) / (hhv - llv) * 100m;

		// store ST values and keep history
		_stQueue.Enqueue(st);
		if (_stQueue.Count > Cycle)
			_stQueue.Dequeue();

		llv = decimal.MaxValue;
		hhv = decimal.MinValue;
		foreach (var value in _stQueue)
		{
			if (value < llv)
				llv = value;
			if (value > hhv)
				hhv = value;
		}

		// final STC calculation
		var stc = hhv == llv ? _prevStc : (st - llv) / (hhv - llv) * 200m - 100m;
		var dStc = stc - _prevStc;
		var color = 4;

		// determine color code
		if (stc > 0)
		{
			if (stc > HighLevel)
				color = dStc >= 0 ? 7 : 6;
			else
				color = dStc >= 0 ? 5 : 4;
		}
		else
		{
			if (stc < LowLevel)
				color = dStc < 0 ? 0 : 1;
			else
				color = dStc < 0 ? 2 : 3;
		}

		// generate trading signals based on color transition
		if (_prevColor.HasValue)
		{
			var prev = _prevColor.Value;

			if (prev > 5)
			{
				if (SellPosClose && Position < 0)
					ClosePosition();

				if (BuyPosOpen && color < 6 && Position <= 0)
					BuyMarket();
			}
			else if (prev < 2)
			{
				if (BuyPosClose && Position > 0)
					ClosePosition();

				if (SellPosOpen && color > 1 && Position >= 0)
					SellMarket();
			}
		}

		_prevColor = color;
		_prevStc = stc;
	}
}

File: API/2015_Color_Schaff_Momentum_Trend_Cycle/PY/color_schaff_momentum_trend_cycle_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import Math, TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import Momentum, DecimalIndicatorValue
from StockSharp.Algo.Strategies import Strategy


class color_schaff_momentum_trend_cycle_strategy(Strategy):

    def __init__(self):
        super(color_schaff_momentum_trend_cycle_strategy, self).__init__()

        self._fast_momentum_length = self.Param("FastMomentum", 23) \
            .SetDisplay("Fast Momentum", "Fast momentum length", "Indicator")
        self._slow_momentum_length = self.Param("SlowMomentum", 50) \
            .SetDisplay("Slow Momentum", "Slow momentum length", "Indicator")
        self._cycle = self.Param("Cycle", 10) \
            .SetDisplay("Cycle", "Cycle length", "Indicator")
        self._high_level = self.Param("HighLevel", 60) \
            .SetDisplay("High Level", "Upper threshold", "Indicator")
        self._low_level = self.Param("LowLevel", -60) \
            .SetDisplay("Low Level", "Lower threshold", "Indicator")
        self._signal_cooldown_bars = self.Param("SignalCooldownBars", 12) \
            .SetDisplay("Signal Cooldown", "Bars to wait between trades", "Trading")
        self._buy_pos_open = self.Param("BuyPosOpen", True) \
            .SetDisplay("Enable Long", "Allow long entries", "Trading")
        self._sell_pos_open = self.Param("SellPosOpen", True) \
            .SetDisplay("Enable Short", "Allow short entries", "Trading")
        self._buy_pos_close = self.Param("BuyPosClose", True) \
            .SetDisplay("Close Long", "Allow closing long positions", "Trading")
        self._sell_pos_close = self.Param("SellPosClose", True) \
            .SetDisplay("Close Short", "Allow closing short positions", "Trading")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Candles timeframe", "General")

        self._macd_history = []
        self._st_history = []
        self._fast_momentum = None
        self._slow_momentum = None
        self._prev_stc = 0.0
        self._prev_color = None
        self._cooldown_remaining = 0

    @property
    def FastMomentum(self):
        return self._fast_momentum_length.Value

    @FastMomentum.setter
    def FastMomentum(self, value):
        self._fast_momentum_length.Value = value

    @property
    def SlowMomentum(self):
        return self._slow_momentum_length.Value

    @SlowMomentum.setter
    def SlowMomentum(self, value):
        self._slow_momentum_length.Value = value

    @property
    def Cycle(self):
        return self._cycle.Value

    @Cycle.setter
    def Cycle(self, value):
        self._cycle.Value = value

    @property
    def HighLevel(self):
        return self._high_level.Value

    @HighLevel.setter
    def HighLevel(self, value):
        self._high_level.Value = value

    @property
    def LowLevel(self):
        return self._low_level.Value

    @LowLevel.setter
    def LowLevel(self, value):
        self._low_level.Value = value

    @property
    def SignalCooldownBars(self):
        return self._signal_cooldown_bars.Value

    @SignalCooldownBars.setter
    def SignalCooldownBars(self, value):
        self._signal_cooldown_bars.Value = value

    @property
    def BuyPosOpen(self):
        return self._buy_pos_open.Value

    @BuyPosOpen.setter
    def BuyPosOpen(self, value):
        self._buy_pos_open.Value = value

    @property
    def SellPosOpen(self):
        return self._sell_pos_open.Value

    @SellPosOpen.setter
    def SellPosOpen(self, value):
        self._sell_pos_open.Value = value

    @property
    def BuyPosClose(self):
        return self._buy_pos_close.Value

    @BuyPosClose.setter
    def BuyPosClose(self, value):
        self._buy_pos_close.Value = value

    @property
    def SellPosClose(self):
        return self._sell_pos_close.Value

    @SellPosClose.setter
    def SellPosClose(self, value):
        self._sell_pos_close.Value = value

    @property
    def CandleType(self):
        return self._candle_type.Value

    @CandleType.setter
    def CandleType(self, value):
        self._candle_type.Value = value

    def OnStarted2(self, time):
        super(color_schaff_momentum_trend_cycle_strategy, self).OnStarted2(time)

        self._fast_momentum = Momentum()
        self._fast_momentum.Length = self.FastMomentum
        self._slow_momentum = Momentum()
        self._slow_momentum.Length = self.SlowMomentum
        self._macd_history = []
        self._st_history = []
        self._prev_stc = 0.0
        self._prev_color = None
        self._cooldown_remaining = 0

        self.SubscribeCandles(self.CandleType) \
            .Bind(self.ProcessCandle) \
            .Start()

    def ProcessCandle(self, candle):
        if candle.State != CandleStates.Finished:
            return

        if self._cooldown_remaining > 0:
            self._cooldown_remaining -= 1

        close = float(candle.ClosePrice)
        fast_input = DecimalIndicatorValue(self._fast_momentum, candle.ClosePrice, candle.OpenTime)
        fast_input.IsFinal = True
        fast_result = self._fast_momentum.Process(fast_input)
        slow_input = DecimalIndicatorValue(self._slow_momentum, candle.ClosePrice, candle.OpenTime)
        slow_input.IsFinal = True
        slow_result = self._slow_momentum.Process(slow_input)
        if not fast_result.IsFormed or not slow_result.IsFormed:
            return

        diff = float(fast_result) - float(slow_result)
        cycle = self.Cycle

        self._add_value(self._macd_history, diff, cycle)
        if len(self._macd_history) < cycle:
            return

        macd_min, macd_max = self._get_min_max(self._macd_history)
        previous_st = self._st_history[-1] if len(self._st_history) > 0 else 0.0
        if macd_max == macd_min:
            st = previous_st
        else:
            st = (diff - macd_min) / (macd_max - macd_min) * 100.0
        self._add_value(self._st_history, st, cycle)

        st_min, st_max = self._get_min_max(self._st_history)
        if st_max == st_min:
            stc = self._prev_stc
        else:
            stc = (st - st_min) / (st_max - st_min) * 200.0 - 100.0
        delta = stc - self._prev_stc
        color = self._get_color(stc, delta)

        if self._prev_color is not None and self._cooldown_remaining == 0:
            if self._prev_color == 6 and color == 7 and self.BuyPosOpen and self.Position <= 0:
                if self.Position < 0 and self.SellPosClose:
                    self.BuyMarket(abs(self.Position))
                self.BuyMarket()
                self._cooldown_remaining = self.SignalCooldownBars
            elif self._prev_color == 1 and color == 0 and self.SellPosOpen and self.Position >= 0:
                if self.Position > 0 and self.BuyPosClose:
                    self.SellMarket(self.Position)
                self.SellMarket()
                self._cooldown_remaining = self.SignalCooldownBars
            elif self.Position > 0 and self.BuyPosClose and color <= 1:
                self.SellMarket(self.Position)
                self._cooldown_remaining = self.SignalCooldownBars
            elif self.Position < 0 and self.SellPosClose and color >= 6:
                self.BuyMarket(abs(self.Position))
                self._cooldown_remaining = self.SignalCooldownBars

        self._prev_color = color
        self._prev_stc = stc

    def _add_value(self, values, value, limit):
        values.append(value)
        if len(values) > limit:
            values.pop(0)

    def _get_min_max(self, values):
        min_val = values[0]
        max_val = values[0]
        for i in range(1, len(values)):
            val = values[i]
            if val < min_val:
                min_val = val
            if val > max_val:
                max_val = val
        return min_val, max_val

    def _get_color(self, stc, delta):
        high = float(self.HighLevel)
        low = float(self.LowLevel)

        if stc > 0:
            if stc > high:
                return 7 if delta >= 0 else 6
            return 5 if delta >= 0 else 4

        if stc < low:
            return 0 if delta < 0 else 1

        return 2 if delta < 0 else 3

    def OnReseted(self):
        super(color_schaff_momentum_trend_cycle_strategy, self).OnReseted()
        self._fast_momentum = None
        self._slow_momentum = None
        self._macd_history = []
        self._st_history = []
        self._prev_stc = 0.0
        self._prev_color = None
        self._cooldown_remaining = 0

    def CreateClone(self):
        return color_schaff_momentum_trend_cycle_strategy()
